### PR TITLE
Fix for unsubscription new subscription race condition

### LINKED ISSUE #4852: Subscription Web Sockets closing prematurely.
**Describe the bug**
I have a React web application utilizing AppSync subscriptions.  Intermittently, when the route/page containing the subscription loads, the subscriptions initiate and then the web socket is immediately closed.  I can see this by turning on DEBUG level logging for the Amplify client.  It seems like the issue happens when page loads are handled via react-router-dom using history.push to navigate to the route (in the Single Page Application).  If you perform a full page refresh in the web browser, the subscriptions seem to load properly.  

Note - the app has 2 distinct subscriptions that are loaded on separate pages/routes.  The first page of the app subscribes to all updates, selecting an item on that page navigates the user to a detail page that subscribes to the updates for that specific id.  I see these web socket closes when you navigate from the summary page to the detail page and also when you navigate from a detail page back to the summary page.  In both cases, navigating via history.push in react-router-dom.  Full page refreshes on either page result in the subscriptions working fine.

I should also note this is a typescript application and authentication is via Cognito User Pools.

**To Reproduce**

1. create a typescript based react web app using React 16.12.0, react-router-dom 5.1.2
2. create an AppSync model and subscriptions using the Amplify CLI (4.13.1)
3. add 2 routes/pages to the web app
4. create a subscription on each page using  aws-amplify: 2.2.2", aws-amplify-react: 3.1.3 and a mechanism to display the values from the subscriptions
6. use react-router-dom to navigate back and forth from each page
7. on one of the pages subscribe to all update events in the model
8. on the second page subscribe to the update events

-- or --
I also have an app on the web that I can point you to that demonstrates the behavior.  I could provide you this link offline.

Link to sample code
Subscriptions are initiated in the useEffect hook on each page:
```
interface ISensorSubscriptionResponse {
  value: {
    data: {
      onUpdateSensor: {
        name: string,
        value: number
      }
    }
  }
}
  const [name, setName] = useState("");
  const [value, setValue] = useState<number | null>(null);

  //subscribe to changes to the sensor's value
  useEffect(() => {  

      console.log('start subscription to sensor');

      const subscriber = API.graphql(graphqlOperation(onUpdateSensor, {id: id})).subscribe({
        next: (response: ISensorSubscriptionResponse) => {
  
          //update the sensor's status in state
          if (response.value.data.onUpdateSensor) {
            setName(response.value.data.onUpdateSensor.name)
            setValue(response.value.data.onUpdateSensor.value)

            console.log('sensor value received');

          }
        },
        error: (error: any) => {
          console.log('error on sensor subscription', error);
        }
      });

      return () => {
        console.log('terminating subscription to sensor');
        subscriber.unsubscribe();
      }
  }, [id]);
```

**Expected behavior**
Expect web socket connection to not terminate immediately on the load of the page/route.

**Screenshots**
Details from the browser console log when Amplify DEBUG logging is enabled:

image

<details>
  <summary><strong>Environment</strong></summary>

```
  System:
    OS: macOS Mojave 10.14.6
    CPU: (8) x64 Intel(R) Core(TM) i7-6920HQ CPU @ 2.90GHz
    Memory: 54.96 MB / 16.00 GB
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 12.14.1 - /usr/local/bin/node
    Yarn: 1.16.0 - /usr/local/bin/yarn
    npm: 6.13.4 - ~/.npm-global/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  Browsers:
    Chrome: 79.0.3945.130
    Firefox: 68.4.1
    Safari: 13.0.5
  npmPackages:
    @material-ui/core: ^4.8.3 => 4.8.3 
    @material-ui/icons: ^4.5.1 => 4.5.1 
    @testing-library/jest-dom: ^4.2.4 => 4.2.4 
    @testing-library/react: ^9.4.0 => 9.4.0 
    @testing-library/user-event: ^7.2.1 => 7.2.1 
    @types/chart.js: ^2.9.11 => 2.9.11 
    @types/jest: ^24.0.25 => 24.0.25 
    @types/node: ^12.12.24 => 12.12.24 
    @types/react: ^16.9.17 => 16.9.17 
    @types/react-chartjs-2: ^2.5.7 => 2.5.7 
    @types/react-dom: ^16.9.4 => 16.9.4 
    @types/react-map-gl: ^5.1.0 => 5.1.0 
    @types/react-router-dom: ^5.1.3 => 5.1.3 
    aws-amplify: ^2.2.2 => 2.2.2 
    aws-amplify-react: ^3.1.3 => 3.1.3 
    chart.js: ^2.9.3 => 2.9.3 
    react: ^16.12.0 => 16.12.0 
    react-chartjs-2: ^2.9.0 => 2.9.0 
    react-dom: ^16.12.0 => 16.12.0 
    react-map-gl: ^5.2.1 => 5.2.1 
    react-router-dom: ^5.1.2 => 5.1.2 
    react-scripts: 3.3.0 => 3.3.0 
    typescript: ^3.7.4 => 3.7.4 
  npmGlobalPackages:
    @aws-amplify/cli: 4.13.1
    license-checker: 25.0.1
    npm: 6.13.4
    oss-attribution-generator: 1.7.1
    serve: 11.2.0
    tldrlegal: 1.0.8
```

</details>

**Smartphone (please complete the following information):**

**Additional context**

**Sample code**
Include additional sample code or a sample repository to help us reproduce the issue. (Be sure to remove any sensitive data)

**_You can turn on the debug mode to provide more info for us by setting window.LOG_LEVEL = 'DEBUG'; in your app._**

### PR DESCRIPTION
**Pull request by elorzafe**
_Issue #, if available:_
Fixes #4852
_Description of changes:_
Fix race condition during unsubscribe and subscribe

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/pubsub/src/Providers/AWSAppSyncRealTimeProvider.ts [hunk 1]
```diff
@@ -372,22 +372,27 @@ export class AWSAppSyncRealTimeProvider extends AbstractPubSubProvider {
...
+		// Verifying 1000ms after removing subscription in case there are new subscription unmount/mount
+		setTimeout(this._closeSocketIfRequired.bind(this), 1000);
 	}
 
-	private _closeSocketWhenFlushed() {
-		logger.debug('closing WebSocket...');
-		clearTimeout(this.keepAliveTimeoutId);
+	private _closeSocketIfRequired() {
+		if (this.subscriptionObserverMap.size > 0) {
+			// No active subscriptions on the WebSocket
```
Comment (elorzafe) [inline packages/pubsub/src/Providers/AWSAppSyncRealTimeProvider.ts]:
```suggestion
			// Active subscriptions on the WebSocket
```

**Comment by houmark** [comment]
Release today @elorzafe @manueliglesias? — can't wait to see if this stabilizes our random non-working subscriptions — this may be related to #4733.

**Comment by elorzafe** [comment]
@houmark 
> Release today @elorzafe @manueliglesias? — can't wait to see if this stabilizes our random non-working subscriptions — this may be related to #4733.

Is already published on @unstable tag on npm

**Comment by houmark** [comment]
> Is already published on @unstable tag on npm

OK, trying that one out :)

**Comment by houmark** [comment]
@elorzafe I mentioned before that some errors in subscriptions are not catchable with the `error` callback. I've had the pleasure of being on a super unstable internet connection the last few days with huge packet loss and a good percentage of subscription connections error out, but my error function does not get the callback.

The console shows the following error when this handskake error happens and it would be nice to be able to react to that in the `error` callback, with the intent to retry connecting:

```javascript
AWSAppSyncRealTimeProvider.ts:620 WebSocket connection to 'wss://retracted.appsync-realtime-api.us-east-1.amazonaws.com/graphql?header=....retracted.....=&payload=e30=' failed: WebSocket opening handshake timed out
(anonymous) @ AWSAppSyncRealTimeProvider.ts:620
(anonymous) @ AWSAppSyncRealTimeProvider.ts:619
(anonymous) @ AWSAppSyncRealTimeProvider.ts:632
step @ AWSAppSyncProvider.ts:204
(anonymous) @ AWSAppSyncProvider.ts:204
(anonymous) @ AWSAppSyncProvider.ts:204
push../node_modules/@aws-amplify/pubsub/lib-esm/Providers/AWSAppSyncRealTimeProvider.js.__awaiter @ AWSAppSyncProvider.ts:204
AWSAppSyncRealTimeProvider._initializeHandshake @ AWSAppSyncRealTimeProvider.ts:613
(anonymous) @ Retry.ts:37
step @ Reachability.ts:25
(anonymous) @ Reachability.ts:25
(anonymous) @ Reachability.ts:25
push../node_modules/@aws-amplify/core/lib-esm/Util/Retry.js.__awaiter @ Reachability.ts:25
retry @ Retry.ts:26
jitteredExponentialRetry @ Retry.ts:78
(anonymous) @ AWSAppSyncRealTimeProvider.ts:606
step @ AWSAppSyncProvider.ts:204
(anonymous) @ AWSAppSyncProvider.ts:204
(anonymous) @ AWSAppSyncProvider.ts:204
push../node_modules/@aws-amplify/pubsub/lib-esm/Providers/AWSAppSyncRealTimeProvider.js.__awaiter @ AWSAppSyncProvider.ts:204
AWSAppSyncRealTimeProvider._initializeRetryableHandshake @ AWSAppSyncRealTimeProvider.ts:604
(anonymous) @ AWSAppSyncRealTimeProvider.ts:580
step @ AWSAppSyncProvider.ts:204
(anonymous) @ AWSAppSyncProvider.ts:204
fulfilled @ AWSAppSyncProvider.ts:204
```
I know this has been merged, and if you like, I can create a new issue with the information above.

And on a stable connection, things seem much more stable, so if it's possible to catch and handle similar issues like the above, I think it's possible for a developer to retry/handle any type of error in a sensible way.

**Pull request merged**